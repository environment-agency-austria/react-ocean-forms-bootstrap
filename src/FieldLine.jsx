/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faSpinner from '@fortawesome/fontawesome-free-solid/faSpinner';
import faExclamationCircle from '@fortawesome/fontawesome-free-solid/faExclamationCircle';
import {
  FormGroup,
  Label,
  Col,
  Button,
  Alert,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap';
import { fieldMetaShape, fieldShape, validators as defaultValidators } from 'react-ocean-forms';

import { BaseFieldError } from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  /**
   * Creates a * marker for the label if the field
   * contains a default required validator
   */
  createRequiredMarker() {
    const {
      validators,
      meta,
    } = this.props;

    if (meta.plaintext) {
      // Hide the required star in plaintext mode
      return null;
    }

    if (Array.isArray(validators) && validators.includes(defaultValidators.required)) {
      const requiredTitle = meta.stringFormatter('ojs_field_required');
      return <span className="field-required" title={requiredTitle}> *</span>;
    }

    return null;
  }

  /**
   * Creates a info button addon if the field
   * has a info property provided
   */
  createInfoAddonButton() {
    const {
      info,
    } = this.props;

    if (info) {
      return (
        <InputGroupAddon addonType="append">
          <Button onClick={this.toggleInfo} outline>
            <FontAwesomeIcon icon={faExclamationCircle} />
          </Button>
        </InputGroupAddon>
      );
    }

    return null;
  }

  createAddon(item, type) {
    if (!item) return null;

    const { meta } = this.props;

    let child = null;
    if (typeof item === 'function') {
      child = item();
    } else {
      child = meta.stringFormatter(item);
    }

    return (
      <InputGroupAddon addonType={type}>
        {child}
      </InputGroupAddon>
    );
  }

  createPrefix() {
    const { prefix } = this.props;
    return this.createAddon(prefix, 'prepend');
  }

  createSuffix() {
    const { suffix } = this.props;
    return this.createAddon(suffix, 'append');
  }

  /**
   * Creates a info alert if the field
   * has a info property provided
   */
  createInfoAlert() {
    const {
      info,
      meta,
    } = this.props;

    const {
      infoVisible,
    } = this.state;

    if (info) {
      const infoString = meta.stringFormatter(info);
      return (
        <Alert
          color="success"
          className="mt-2"
          isOpen={infoVisible}
          toggle={this.toggleInfo}
        >
          {infoString}
        </Alert>
      );
    }

    return null;
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
    } = this.props;

    // Generate optional elements
    const requiredMarker = this.createRequiredMarker();
    const infoButton = this.createInfoAddonButton();
    const infoAlert = this.createInfoAlert();

    const invalidAlert = meta.valid
      ? null
      : <FontAwesomeIcon icon={faExclamationCircle} className="ico-invalid mr-2" />;

    const validatingSpinner = meta.isValidating
      ? <FontAwesomeIcon icon={faSpinner} spin className="ico-loading mr-2" />
      : null;

    const groupClass = meta.valid ? '' : 'is-invalid';
    const labelString = meta.stringFormatter(label);

    const prefixItem = this.createPrefix();
    const suffixItem = this.createSuffix();

    return (
      <FormGroup row className={groupClass}>
        <Label sm="3" for={field.id} className="text-right">
          {invalidAlert}
          {validatingSpinner}
          {labelString}
          {requiredMarker}
        </Label>
        <Col sm="9">
          <InputGroup>
            {prefixItem}
            {children}
            {suffixItem}
            {infoButton}
            <BaseFieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
              context={{ stringFormatter: meta.stringFormatter }}
            />
          </InputGroup>

          {infoAlert}
        </Col>
      </FormGroup>
    );
  }
}

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
};

export default FieldLine;
