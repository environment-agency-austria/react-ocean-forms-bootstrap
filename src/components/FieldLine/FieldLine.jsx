/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
