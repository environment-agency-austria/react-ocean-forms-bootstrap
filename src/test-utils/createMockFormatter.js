export default () => jest.fn().mockImplementation(id => id);
