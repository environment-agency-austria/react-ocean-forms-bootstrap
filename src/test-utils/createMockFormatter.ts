export const createMockFormatter = (): jest.Mock => jest.fn().mockImplementation((id: string) => id);
