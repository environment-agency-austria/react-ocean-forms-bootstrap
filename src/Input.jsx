/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms';

import FieldLine from './FieldLine';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  const invalid = meta.valid === true ? undefined : true;

  return (
    <FieldLine {...props}>
      <StrapInput type={type} {...field} invalid={invalid} />
    </FieldLine>
  );
}

Input.defaultProps = {
  info: undefined,
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
