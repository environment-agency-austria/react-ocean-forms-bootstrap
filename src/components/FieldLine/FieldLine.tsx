/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText, PropsOf } from 'react-ocean-forms';
import { Col, InputGroup, Label } from 'reactstrap';

import { FieldError } from './FieldError';
import { FieldLineAddon } from './FieldLineAddon';
import { FieldRow } from './FieldRow';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import { InvalidAlert } from './InvalidAlert';
import { RequiredMarker } from './RequiredMarker';
import { ValidatingSpinner } from './ValidatingSpinner';

import { IFieldLineProps } from './FieldLine.types';

interface IFieldLineState {
  infoVisible: boolean;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export class FieldLine extends React.Component<IFieldLineProps, IFieldLineState> {
  public static displayName: string = 'FieldLine';

  public static defaultProps = {
    labelSize: '3',
    inputSize: '9',
    labelClass: 'text-right',
  };

  constructor(props: IFieldLineProps) {
    super(props);

    this.state = {
      infoVisible: false,
    };
  }

  /**
   * Toggles the visibility of the info alert
   */
  private toggleInfo = (): void => {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  public render(): JSX.Element {
    const {
      field,
      meta,
      label,
      children,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

export type DefaultizedFieldLineProps = JSX.LibraryManagedAttributes<typeof FieldLine, PropsOf<typeof FieldLine>>;
