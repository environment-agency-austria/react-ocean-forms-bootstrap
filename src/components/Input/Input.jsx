/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldLine } from '../FieldLine';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  const invalid = meta.valid === true ? undefined : true;

  return (
    <FieldLine {...props}>
      <StrapInput type={type} {...field} invalid={invalid} plaintext={meta.plaintext}>
        {meta.plaintext ? field.value : undefined}
      </StrapInput>
    </FieldLine>
  );
}

Input.displayName = 'Input';

Input.defaultProps = {
  info: undefined,
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
