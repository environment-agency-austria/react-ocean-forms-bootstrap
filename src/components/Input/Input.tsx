/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { useField } from 'react-ocean-forms';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { IInputProps } from './Input.types';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * react-ocean-forms support
 */
export const Input = <TSubmitValue extends unknown>(props: IInputProps<TSubmitValue>): JSX.Element => {
  const {
    type = 'text',
    ...rest
  } = props;

  const { fieldProps, metaProps } = useField(rest);

  const invalid = metaProps.valid === true ? undefined : true;
  const defaultizedValue = fieldProps.value ?? '';

  return (
    <FieldLine {...rest} fieldProps={fieldProps} metaProps={metaProps}>
      <StrapInput type={type} {...fieldProps} value={defaultizedValue} invalid={invalid} plaintext={metaProps.plaintext} />
    </FieldLine>
  );
};
