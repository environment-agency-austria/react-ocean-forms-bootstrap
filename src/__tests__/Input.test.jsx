import React from 'react';
import { shallow } from 'enzyme';

import createMockFormatter from '../test-utils/createMockFormatter';
import Input from '../Input';
import FieldLine from '../FieldLine';

describe('<Input />', () => {
  const FIELD_ID = 'field0';
  const FIELD_NAME = 'field0';
  const FIELD_LABEL = 'field0';

  const meta = {
    valid: true,
    error: undefined,
    isValidating: undefined,
    stringFormatter: createMockFormatter(),
  };
  const field = {
    value: '',
    invalid: false,
    id: FIELD_ID,
    name: FIELD_NAME,
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  const setup = props => shallow((
    <Input
      label={FIELD_LABEL}
      meta={meta}
      field={field}
      {...props}
    />
  ));

  afterEach(() => {
    field.onChange.mockClear();
    field.onBlur.mockClear();
  });

  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a field line and a bootstrap input', () => {
    const wrapper = setup();

    expect(wrapper.find(FieldLine).exists()).toBeTruthy();
    expect(wrapper.find('Input').exists()).toBeTruthy();
  });

  it('should render a text input by default', () => {
    const wrapper = setup();

    expect(wrapper.find('Input').prop('type')).toEqual('text');
  });

  it('should let the user override the input type', () => {
    const INPUT_TYPE = 'number';
    const wrapper = setup({ type: INPUT_TYPE });

    expect(wrapper.find('Input').prop('type')).toEqual(INPUT_TYPE);
  });

  it('should pass the meta props to the field line', () => {
    const wrapper = setup();

    expect(wrapper.find(FieldLine).prop('meta')).toEqual(meta);
  });

  it('should call field.onChange when the input changes', () => {
    const wrapper = setup();
    const event = { target: { name: FIELD_ID, value: 'test_value' } };

    wrapper.find('Input').simulate('change', event);
    expect(field.onChange).toHaveBeenCalledWith(event);
  });

  it('should call field.onBlur when there is an input blur', () => {
    const wrapper = setup();
    const event = { target: { name: FIELD_ID } };

    wrapper.find('Input').simulate('blur', event);
    expect(field.onBlur).toHaveBeenCalledWith(event);
  });
});
