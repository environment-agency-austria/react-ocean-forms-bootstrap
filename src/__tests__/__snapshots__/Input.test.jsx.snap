// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    field={
      Object {
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": undefined,
        "isValidating": undefined,
        "stringFormatter": [MockFunction],
        "valid": true,
      }
    }
    type="text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Input
        id="field0"
        name="field0"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value=""
      />,
      "field": Object {
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "info": undefined,
      "label": "field0",
      "meta": Object {
        "error": undefined,
        "isValidating": undefined,
        "stringFormatter": [MockFunction],
        "valid": true,
      },
      "prefix": undefined,
      "suffix": undefined,
      "type": "text",
      "validators": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "id": "field0",
        "invalid": undefined,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Input
          id="field0"
          name="field0"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          type="text"
          value=""
        />,
        "field": Object {
          "id": "field0",
          "invalid": false,
          "name": "field0",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "info": undefined,
        "label": "field0",
        "meta": Object {
          "error": undefined,
          "isValidating": undefined,
          "stringFormatter": [MockFunction],
          "valid": true,
        },
        "prefix": undefined,
        "suffix": undefined,
        "type": "text",
        "validators": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "field0",
          "invalid": undefined,
          "name": "field0",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
